src/jit/valuenum.cpp

```diff
@@ -1613,6 +1613,7 @@ double ValueNumStore::GetConstantDouble(ValueNum argVN)
 {
     assert(IsVNConstant(argVN));
     var_types argVNtyp = TypeOfVN(argVN);
+    assert(varTypeIsFloating(argVNtyp));
```

Comment:
> Yeah... in fact, it looks like we're already (without this change) constant-folding the single-precision intrinsics, just doing it by widening to double and evaluating and truncating back down, is that right?

Yes, which can result in lower performance and different computed results for the new System.MathF functions.

> The language in the spec certainly seems to give us the liberty to do that (I.12.1.3).

I believe that most consumers of the single-precision functions would rather have the speed than the additional precision (including any additional precision retained by upcasting, computing, and downcasting).

> I'd think we'd want to minimize differences between optimized and unoptimized code's results

I agree, and my understanding is that 'internal representation' is a platform agnostic way of saying whatever form the CPU accepts (so 80-bits when using the x87 FPU, for example). I think that using the C type names (float and double), calling the appropriate versions of the functions (`sqrtf` and `sqrt`, respectively), and then letting the native runtime handle the internal implementation details is the best way to go here.